README.md
# screeps-power-planner
Power Planner for your Power Creeps


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## v0.6 -- UI touch-ups
### [214de55](https://github.com/OrletSoir/screeps-power-planner/commit/214de55d17962b8fb8a151111c4b6b04d9ff187a) -- 2019-05-25
* Added "undo" button to every power box for ease of access.

## v0.5 -- Feature-complete release
### [28bef5e](https://github.com/OrletSoir/screeps-power-planner/commit/28bef5e651550b5f90f0df70090d71c309b314aa) -- 2019-04-17
* fixed small error in `index.php`
* updated powers data with data pulled directly from client

### [b68c876](https://github.com/OrletSoir/screeps-power-planner/commit/b68c87696e992f507d4bb0af26c641d23bdd2672) -- 2019-04-14
* sharing via link
* parsing of build array
* reset button

## v0.1 -- Bare bones version
### [3de1d9e](https://github.com/OrletSoir/screeps-power-planner/commit/3de1d9e3d08f46a6a349678ca0d9fcf63d1d20e7) -- 2019-04-07
* clicking the "Pick Order" will give a nice copyable box of power array
* Undo button! That actually undoes last step!

### [1f10361](https://github.com/OrletSoir/screeps-power-planner/commit/1f10361f5f6559a90f1de075ea6d1c0f79ad4c0c) -- 2019-04-01
* power upgrade buttons actually do stuff
* changed zero lvl badge color
* adding skills to skill list
* adding/updating/sorting selected skill list

### [93112f6](https://github.com/OrletSoir/screeps-power-planner/commit/93112f672b392988231db2ba809dd0ef28980e44) -- 2019-04-31
* It's a v0.1!

### [e0b40af](https://github.com/OrletSoir/screeps-power-planner/commit/e0b40af8af7f69f854c2198feec65f41e54ae92e) -- 2019-04-31
* Initial commit


power-planner/index.php
<?php

/**
 * @author Orlet Soir
 * @copyright 2019
 * 
 * Screeps Power Planner for Power Creeps
 */

define('VERSION_STRING', "v0.6");

$powersParam = "";
$valid = true;
if (isset($_REQUEST['p']) && strlen($_REQUEST['p']) > 0)
{
	$parts = explode(",", $_REQUEST['p']);
	
	foreach ($parts as $part)
	{
		if (!is_numeric($part))
		{
			$valid = false;
			break;
		}
	}
	
	if ($valid)
		$powersParam = $_REQUEST['p'];
}

?>
<!doctype html>
<html class="no-js" lang="en" dir="ltr">
	<head>
		<meta charset="utf-8">
		<meta http-equiv="x-ua-compatible" content="ie=edge">
		<meta name="viewport" content="width=device-width, initial-scale=1.0">
		<title>Screeps Power Planner</title>
		<link rel="stylesheet" href="css/foundation.css">
		<link rel="stylesheet" href="css/app.css">
		<script language="javascript">var powersParam = "<?php echo $powersParam;?>";</script>
	</head>
	<body>
		&nbsp;
		<div class="grid-container">
			<!-- page header -->
			<div class="grid-x grid-padding-x">
				<div class="large-24 cell text-center">
					<div class="callout">
						<h4>Screeps Power Planner <small><?php echo VERSION_STRING; ?></small></h4>
					</div>
				</div>
			</div>
<?php
	if (!$valid)
	{
?>
			<!-- invalid parameters alert -->
			<div class="callout alert" data-closable>
				<h5>Invalid parameters</h5>
				<p>The url contains invalid characters and could not be parsed properly.</p>
				<button class="close-button" aria-label="Dismiss alert" type="button" data-close>
					<span aria-hidden="true">&times;</span>
				</button>
			</div>
<?php
	}
?>
			<!-- main body container -->
			<div class="grid-x grid-padding-x" id="duo-pane-eq" data-equalizer="foo">
				<div id="powers-available-container" class="large-18 cell">
					<div class="callout" data-equalizer-watch="foo">
						<label>Class:</label>
						<select id="class" required onchange="onClassChange();" disabled="DISABLED">
    						<option disabled="DISABLED" selected="SELECTED" value="operator" class="disabled-option">Operator</option>
						</select>
						<div class="grid-x grid-padding-x" id="powers-wrapper" data-equalizer="bar" data-equalize-by-row="true">
							&nbsp;
						</div>
					</div>
				</div>
				<div id="powers-selected-container" class="large-6 cell text-center">
					<div class="callout" data-equalizer-watch="foo">
						<h6 class="power-creep-info">
							Lv.: <small id="display-level">0</small> / Hits: <small id="display-hits">1000</small> / Carry: <small id="display-carry">100</small>
						</h6>
						<hr />
						<h6><u>Selected Powers</u></h6>
						<div id="powers-selected-list"></div>
						<hr />
						<h6 data-toggle="powers-copy-list"><u>Pick Order</u></h6>
						<div id="powers-copy-list" class="dropdown-pane" data-dropdown data-position="bottom" data-alignment="center" data-auto-focus="true">
							<textarea id="powers-copy-area" onfocus="this.select();" rows="4">[]</textarea>
							<button onclick="pickOrderUpdate();" class="button small float-right">Parse</button>
						</div>
						<div id="powers-history-list"></div>
						<hr />
						<div>
							<button onclick="resetPowerProgress();" class="button small float-left warning">&#x1F503; Reset</button>
							<button onclick="updateShareLink();" class="button small float-center success">&#x1F517; Share</button>
							<button onclick="undoLevelPower();" class="button small float-right disabled" id="power-undo-static">&#x293E; Undo</button>
						</div>
					</div>
				</div>
			</div>
			<!-- page footer -->
			<div class="grid-x grid-padding-x">
				<div class="large-24 cell text-center">
					<div class="callout page-footer text-center">
						| Screeps Power Planner <?php echo VERSION_STRING; ?> by Orlet | <a href="https://github.com/OrletSoir/screeps-power-planner">GitHub repo</a> | <a href="https://screeps.com/">Screeps</a> &ndash; The world's first MMO sandbox game for programmers | 
					</div>
				</div>
			</div>
		</div>
		
		<script src="js/vendor/jquery.js"></script>
		<script src="js/vendor/what-input.js"></script>
		<script src="js/vendor/foundation.js"></script>
		<script src="js/constants.js"></script>
		<script src="js/app.js"></script>
	</body>
</html>


power-planner/js/app.js
// formatting function, stolen from stackexchange
String.prototype.formatUnicorn = String.prototype.formatUnicorn || function () {
    "use strict";
    var str = this.toString();
    if (arguments.length) {
        var t = typeof arguments[0];
        var key;
        var args = ("string" === t || "number" === t) ?
            Array.prototype.slice.call(arguments)
            : arguments[0];

        for (key in args) {
            str = str.replace(new RegExp("\\{" + key + "\\}", "gi"), args[key]);
        }
    }

    return str;
};

// sub-element sorting function
function sortUsingNestedText(parent, childSelector, keySelector) {
    var items = parent.children(childSelector).sort(function(a, b) {
        var vA = $(keySelector, a).text();
        var vB = $(keySelector, b).text();
        return (vA < vB) ? -1 : (vA > vB) ? 1 : 0;
    });
    parent.append(items);
};

function sortReverseUsingNestedText(parent, childSelector, keySelector) {
    var items = parent.children(childSelector).sort(function(a, b) {
        var vA = $(keySelector, a).text();
        var vB = $(keySelector, b).text();
        return (vA < vB) ? 1 : (vA > vB) ? -1 : 0;
    });
    parent.append(items);
};

// globals
var creepClass = "";
var creepLevel = 0;
var powerHistory = [];

// callback to populate powers for selected class 
function populatePowers(powerClass)
{
	$("#powers-wrapper").empty();
	$("#powers-selected-list").empty();
	$("#powers-history-list").empty();
	
	// save values
	creepClass = powerClass;
	creepLevel = 0;
	
	// re-populate powers
	for (var power in POWER_INFO)
	{
		if (!POWER_INFO[power] || !POWER_INFO[power].className || POWER_INFO[power].className != powerClass)
			continue;
		
		addPower(power, POWER_INFO[power], POWER_DESCRIPTIONS[power]);
	}
	
	// reflow the new stuff
	Foundation.reInit('equalizer');	
	$(document).foundation();	
}

// return either value or glued value if it's an array
function formatMultiValue(value)
{
	if (!value.length || !value.join)
		return value;
	
	return value.join("/");
}

// format power description text
function formatPowerEffect(id, info, description)
{
	// h4x
	id = parseInt(id);
	var mag = ["", "K", "M", "B", "T"];
	
	switch (id)
	{
		// handle special cases
		case PWR_OPERATE_SPAWN:
		case PWR_OPERATE_TERMINAL:
		case PWR_DISRUPT_TOWER:
			var effect = [];
			for (var i = 0; i < info.effect.length; i++)
				effect.push(Math.round((1 - info.effect[i])*100)+"%");
				
			return description.formatUnicorn(formatMultiValue(effect));
		
		case PWR_OPERATE_TOWER:
			var effect = [];
			for (var i = 0; i < info.effect.length; i++)
				effect.push(Math.round((info.effect[i] - 1)*100)+"%");
				
			return description.formatUnicorn(formatMultiValue(effect));
		
		case PWR_OPERATE_EXTENSION:
			var effect = [];
			for (var i = 0; i < info.effect.length; i++)
				effect.push(Math.round((info.effect[i])*100)+"%");
				
			return description.formatUnicorn(formatMultiValue(effect));
		
		case PWR_SHIELD:
		case PWR_OPERATE_STORAGE:
			var effect = [];
			for (var i = 0; i < info.effect.length; i++)
			{
				var value = info.effect[i];
				var order = 0;
				
				while (value > 1000)
				{
					value = value/1000;
					order++;
				}
				
				effect.push(value+mag[order]);
			}
				
			return description.formatUnicorn(formatMultiValue(effect));
		
		// default formatting shold work for most powers
		default:
			return description.formatUnicorn(info.effect && formatMultiValue(info.effect));
	}
}

// format powers' description (return inner element)
function generateTooltipText(info)
{
	var lines = [];
	
	if (info.ops)
		lines.push("Cost: "+formatMultiValue(info.ops)+" ops");
	
	if (info.range)
		lines.push("Range: "+formatMultiValue(info.range));
	
	if (info.duration)
		lines.push("Duration: "+formatMultiValue(info.duration));
	
	if (info.cooldown)
		lines.push("Cooldown: "+formatMultiValue(info.cooldown));
	
	return $("<p/>").addClass("text-left").append(lines.join("<br/>"));
}

// add the power to powers panel
function addPower(id, info, description)
{
	// main element wrappers
	var div = $("<div/>").addClass("small-12 medium-8 large-8 cell text-center power-container");
	var subdiv = $("<div/>").addClass("callout small").attr("data-equalizer-watch", "bar").attr("data-toggle", "power-info-"+id);
	
	// name and description
	var badge = $("<span>0</span>").attr("id", "power-badge-"+id).addClass("badge secondary float-right");
	
	subdiv.append($("<h6/>").addClass("power-name").append(description.name).append(badge).append($("<br/>")).append($("<small/>").append(formatPowerEffect(id, info, description.text))));
	
	// levels tag
	var levels = $("<div/>").addClass("levels").append(formatMultiValue(info.level));
	
	// buttons aligned to bottom
	var undo = $("<button onclick=\"undoLevelPower();\" />").addClass("button tiny float-left").attr("id", "power-undo-"+id).append("&#x293E;");
	var button = $("<button onclick=\"levelPower("+id+");\" />").addClass("button tiny float-right").attr("id", "power-button-"+id).append("+");
	
	if (creepLevel >= POWER_CREEP_MAX_LEVEL)
		button.addClass("secondary disabled");
	else if (info.level[0] > creepLevel)
		button.addClass("alert disabled");
	else
		button.addClass("success");
	
	if (creepLevel == 0)
		undo.addClass("disabled");
	
	subdiv.append($("<div/>").addClass("button-spacer").append("&nbsp;"));
	subdiv.append($("<div/>").addClass("button-container align-to-bottom power-level-add").append(levels).append(undo).append(button));
	
	div.append(subdiv);
	
	// pop-out detailed info panel
	div.append($("<div/>").addClass("dropdown-pane").attr("id", "power-info-"+id).attr("data-position", "bottom").attr("data-alignment", "center").attr("data-dropdown", "").attr("data-hover", "true").attr("data-hover-pane", "true").append(generateTooltipText(info)));
	
	// put the whole thing in
	$("#powers-wrapper").append(div);
}

// level the power up and update all the things
function levelPower(id)
{
	var powerBlock = $("#power-info-"+id);
	if (!powerBlock.length)
		return;
	
	var powerBadge = $("#power-badge-"+id);
	if (!powerBadge.length)
		return;
	
	var powerInfo = POWER_INFO[id];
	if (!powerInfo)
		return;
	
	// level checks
	var currentLevel = parseInt(powerBadge.text());
	var nextLevel = currentLevel+1;
	
	if (currentLevel > powerInfo.level.length || powerInfo.level[currentLevel] > creepLevel)
		return;
	
	// update the badge
	updatePowerBadge(powerBadge, nextLevel, powerInfo.level.length);
	
	// save to history
	powerHistory.push(id);
	
	// update the internal counters and power button availability
	creepLevel++;
	updateCreepStats();
	updatePowerUpgradeButtons();
	
	// add power to sidebar and update the overall list
	addUpdatePowerToSidebarList(id, nextLevel);
	addPowerToSidebarHistory(id, nextLevel);
	updatePowersCopyArea();
	updatePowerUndoButtons();
}

// add selected power to the sidebar
function addPowerToSidebarHistory(id, level)
{
	var powerSidebar = $("#powers-history-list");
	if (!powerSidebar.length)
		return;
	
	var desc = POWER_DESCRIPTIONS[id];
	if (!desc)
		return;
	
	powerSidebar.append($("<h6/>").addClass("power-item").attr("id", "powers-history-item-"+creepLevel).append(desc.name).append(" ").append($("<small/>").append("Lv. "+level)));
}

// remove selected power from the sidebar
function removePowerFromSidebarHistory(level)
{
	var powerItem = $("#powers-history-item-"+level);
	if (!powerItem.length)
		return;
	
	powerItem.remove();
}

function addUpdatePowerToSidebarList(id, level)
{
	var powerSelectedList = $("#powers-selected-list");
	if (!powerSelectedList.length)
		return;
	
	var desc = POWER_DESCRIPTIONS[id];
	if (!desc)
		return;
	
	var powersItem = $("#powers-selected-item-"+id);
	if (powersItem.length)
	{
		if (level > 0)
		{
			var powersItemLevel = $("small", powersItem);
			
			powersItemLevel.empty();
			powersItemLevel.append("Lv. "+level);
		}
		else
			powersItem.remove();
	}
	else
		powerSelectedList.append($("<h6/>").addClass("power-item").attr("id", "powers-selected-item-"+id).append(desc.name).append(" ").append($("<small/>").append("Lv. "+level)));
	
	sortReverseUsingNestedText(powerSelectedList, "h6", "small");
}

function updatePowersCopyArea()
{
	$("#powers-copy-area").val(JSON.stringify(powerHistory));
}

function addMoveUndoButton()
{
	var button = $("#power-undo-button");
	
	if (button.length < 1)
		button = $("<div/>").attr("id", "power-undo-button").append("&nbsp;").append($("<button onclick=\"undoLevelPower();\" />").addClass("button tiny float-right").append("&#x293E;"));
	
	$("#powers-history-list").append(button);
}

function updatePowerUndoButtons()
{
	var buttons = $("[id^=power-undo-]");
	
	if (buttons.length < 1)
		return;
	
	buttons.removeClass("disabled");
	
	if (creepLevel < 1)
		buttons.addClass("disabled");
}

function removeUndoButton()
{
	var button = $("#power-undo-button");
	
	if (button.length)
		button.remove();
}

function undoLevelPower()
{
	// what's on history
	var lastPower = powerHistory.pop();
	
	var powerBadge = $("#power-badge-"+lastPower);
	if (!powerBadge.length)
		return;
	
	var powerInfo = POWER_INFO[lastPower];
	if (!powerInfo)
		return;
	
	// level checks
	var currentLevel = parseInt(powerBadge.text());
	var nextLevel = currentLevel-1;
	
	// update the badge
	updatePowerBadge(powerBadge, nextLevel, powerInfo.level.length);
	
	// remove power from sidebar and update overall list
	removePowerFromSidebarHistory(creepLevel);
	addUpdatePowerToSidebarList(lastPower, nextLevel);
	
	// update the internal counters and power button availability
	creepLevel--;
	updateCreepStats();
	updatePowerUpgradeButtons();
	
	// add power to sidebar and update the overall list
	updatePowersCopyArea();
	updatePowerUndoButtons();
}

// update the badges for creep level and stats
function updateCreepStats()
{
	var levelDisplay = $("#display-level");
	if (!levelDisplay)
		return;
	
	var hitsDisplay = $("#display-hits");
	if (!hitsDisplay)
		return;
	
	var carryDisplay = $("#display-carry");
	if (!carryDisplay)
		return;
	
	levelDisplay.empty();
	levelDisplay.append(creepLevel);
	
	hitsDisplay.empty();
	hitsDisplay.append((creepLevel+1)*1000);
	
	carryDisplay.empty();
	carryDisplay.append((creepLevel+1)*100);
	
}

// update the power upgrade buttons
function updatePowerUpgradeButtons()
{
	for (var power in POWER_INFO)
	{
		if (!POWER_INFO[power] || !POWER_INFO[power].className || POWER_INFO[power].className != creepClass)
			continue;
		
		enableDisableUpgradeButton(power, POWER_INFO[power]);
	}
}

// update the power badge
function updatePowerBadge(powerBadge, nextLevel, maxLevel)
{
	powerBadge.empty();
	powerBadge.append(nextLevel);
	powerBadge.removeClass("success warning secondary");
	
	if (nextLevel < 1)
		powerBadge.addClass("secondary");
	else if (nextLevel >= maxLevel)
		powerBadge.addClass("warning");
	else
		powerBadge.addClass("success");
}

// enable/disable the power upgrade button for power id
function enableDisableUpgradeButton(id, info)
{
	var powerButton = $("#power-button-"+id);
	if (!powerButton.length)
		return;
	
	var powerBadge = $("#power-badge-"+id);
	if (!powerBadge.length)
		return;
	
	var nextLevel = parseInt(powerBadge.text());
	
	powerButton.removeClass("success alert secondary disabled");
	
	if (creepLevel >= POWER_CREEP_MAX_LEVEL || nextLevel >= info.level.length)
		powerButton.addClass("secondary disabled");
	else if (info.level[nextLevel] > creepLevel)
		powerButton.addClass("alert disabled");
	else
		powerButton.addClass("success");
}

// reset all current selection data
function resetPowerProgress()
{
	// reset creep's levels
	creepLevel = 0;
	powerHistory.length = 0;
	
	$("#powers-history-list").empty();
	$("#powers-selected-list").empty();
	
	// re-populate powers
	for (var power in POWER_INFO)
	{
		if (!POWER_INFO[power] || !POWER_INFO[power].className || POWER_INFO[power].className != creepClass)
			continue;
	
		var powerBadge = $("#power-badge-"+power);
		if (!powerBadge.length)
			continue;
		
		// update the badge
		updatePowerBadge(powerBadge, 0, POWER_INFO[power].level.length);
		enableDisableUpgradeButton(power, POWER_INFO[power]);
	}
	
	updateCreepStats();
	updatePowersCopyArea();
	updatePowerUndoButtons();
}

// process specific power
function processPowerUpgrade(power)
{
	levelPower(power);
}

// process power upgrade path
function processPowersUpgradeArray(powers)
{
	// reset all data
	resetPowerProgress();
	
	// process powers
	for (var i = 0; i < powers.length; i++)
		processPowerUpgrade(powers[i]);
}

// pick order field changed
function pickOrderUpdate()
{
	var powerArea = $("#powers-copy-area");
	if (!powerArea.length)
		return;
	
	var powers = JSON.parse(powerArea.val());
	
	if (powers)
		processPowersUpgradeArray(powers);
}

// set up powers from saved parameter
function processPowersParameter(parameter)
{
	var powers = parameter.split && parameter.split(",");
	
	if (!powers || !powers.length)
		return;
	
	for (var i = 0; i < powers.length; i++)
		processPowerUpgrade(parseInt(powers[i]));
}

// generate share link
function updateShareLink()
{
	window.history.replaceState(powerHistory, 'Powers Share', '?p='+powerHistory);
}

// initialize stuff
$(document).foundation();
$(window).on('load', function()
{
	populatePowers("operator");
	
	if (powersParam && powersParam.length)
		processPowersParameter(powersParam);
});

power-planner/js/constants.js
/**
 * All power creep related constants
 */

// max level
const POWER_CREEP_MAX_LEVEL = 25;

// classes
const POWER_CLASS = {
    OPERATOR: 'operator'
};

// powers
const PWR_GENERATE_OPS = 1;
const PWR_OPERATE_SPAWN = 2;
const PWR_OPERATE_TOWER = 3;
const PWR_OPERATE_STORAGE = 4;
const PWR_OPERATE_LAB = 5;
const PWR_OPERATE_EXTENSION = 6;
const PWR_OPERATE_OBSERVER = 7;
const PWR_OPERATE_TERMINAL = 8;
const PWR_DISRUPT_SPAWN = 9;
const PWR_DISRUPT_TOWER = 10;
const PWR_DISRUPT_SOURCE = 11;
const PWR_SHIELD = 12;
const PWR_REGEN_SOURCE = 13;
const PWR_REGEN_MINERAL = 14;
const PWR_DISRUPT_TERMINAL = 15;
const PWR_OPERATE_POWER = 16;
const PWR_FORTIFY = 17;
const PWR_OPERATE_CONTROLLER = 18;
const PWR_OPERATE_FACTORY = 19;

// powers descriptions
const POWER_DESCRIPTIONS = {
	[PWR_GENERATE_OPS]: {
		name: "GENERATE_OPS",
		text: "Generate {0} ops resource units."
	},
	[PWR_OPERATE_SPAWN]: {
		name: "OPERATE_SPAWN",
		text: "Reduce spawn time by {0}."
	},
	[PWR_OPERATE_TOWER]: {
		name: "OPERATE_TOWER",
		text: "Increase damage, repair and heal amount by {0}."
	},
	[PWR_OPERATE_STORAGE]: {
		name: "OPERATE_STORAGE",
		text: "Increase capacity by {0} units."
	},
	[PWR_OPERATE_LAB]: {
		name: "OPERATE_LAB",
		text: "Increase reaction amount by {0} units."
	},
	[PWR_OPERATE_EXTENSION]: {
		name: "OPERATE_EXTENSION",
		text: "Instantly fill {0} of all extensions in the room using energy from the target structure (container, storage, or terminal)."
	},
	[PWR_OPERATE_OBSERVER]: {
		name: "OPERATE_OBSERVER",
		text: "Grant unlimited range."
	},
	[PWR_OPERATE_TERMINAL]: {
		name: "OPERATE_TERMINAL",
		text: "Decrease transfer energy cost and cooldown by {0}."
	},
	[PWR_DISRUPT_SPAWN]: {
		name: "DISRUPT_SPAWN",
		text: "Pause spawning process."
	},
	[PWR_DISRUPT_TOWER]: {
		name: "DISRUPT_TOWER",
		text: "Reduce effectiveness by {0}."
	},
	[PWR_DISRUPT_SOURCE]: {
		name: "DISRUPT_SOURCE",
		text: "Pause energy regeneration."
	},
	[PWR_SHIELD]: {
		name: "SHIELD",
		text: "Create a temporary non-repairable rampart structure on the same square with {0} hits. Cannot be used on top of another rampart."
	},
	[PWR_REGEN_SOURCE]: {
		name: "REGEN_SOURCE",
		text: "Regenerate {0} energy units in a source every 15 ticks."
	},
	[PWR_REGEN_MINERAL]: {
		name: "REGEN_MINERAL",
		text: "Regenerate {0} mineral units in a deposit every 10 ticks."
	},
	[PWR_DISRUPT_TERMINAL]: {
		name: "DISRUPT_TERMINAL",
		text: "Block withdrawing resources from the terminal."
	},
	[PWR_FORTIFY]: {
		name: "FORTIFY",
		text: "Make a wall or rampart tile invulnerable to all creep attacks and powers."
	},
	[PWR_OPERATE_POWER]: {
		name: "OPERATE_POWER",
		text: "Increase power processing speed of a Power Spawn by {0} units per tick."
	},
	[PWR_OPERATE_CONTROLLER]: {
		name: "OPERATE_CONTROLLER",
		text: "Increase max limit of energy that can be used for upgrading a Level 8 Controller each tick by {0} energy units."
	},
	[PWR_OPERATE_FACTORY]: {
		name: "OPERATE_FACTORY",
		text: "An unknown power."
	},
};

// powers info
const POWER_INFO = {"1":{"className":"operator","level":[0,2,7,14,22],"cooldown":50,"effect":[1,2,4,6,8]},"2":{"className":"operator","level":[0,2,7,14,22],"cooldown":300,"duration":1000,"range":3,"ops":100,"effect":[0.9,0.7,0.5,0.35,0.2]},"3":{"className":"operator","level":[0,2,7,14,22],"cooldown":10,"duration":100,"range":3,"ops":10,"effect":[1.1,1.2,1.3,1.4,1.5]},"4":{"className":"operator","level":[0,2,7,14,22],"cooldown":800,"duration":1000,"range":3,"ops":100,"effect":[500000,1000000,2000000,4000000,7000000]},"5":{"className":"operator","level":[0,2,7,14,22],"cooldown":50,"duration":1000,"range":3,"ops":10,"effect":[2,4,6,8,10]},"6":{"className":"operator","level":[0,2,7,14,22],"cooldown":50,"range":3,"ops":2,"effect":[0.2,0.4,0.6,0.8,1]},"7":{"className":"operator","level":[0,2,7,14,22],"cooldown":400,"duration":[200,400,600,800,1000],"range":3,"ops":10},"8":{"className":"operator","level":[0,2,7,14,22],"cooldown":500,"duration":1000,"range":3,"ops":100,"effect":[0.9,0.8,0.7,0.6,0.5]},"9":{"className":"operator","level":[0,2,7,14,22],"cooldown":5,"range":20,"ops":10,"duration":[1,2,3,4,5]},"10":{"className":"operator","level":[0,2,7,14,22],"cooldown":0,"duration":5,"range":50,"ops":10,"effect":[0.9,0.8,0.7,0.6,0.5]},"11":{"className":"operator","level":[0,2,7,14,22],"cooldown":100,"range":3,"ops":100,"duration":[100,200,300,400,500]},"12":{"className":"operator","level":[0,2,7,14,22],"effect":[5000,10000,15000,20000,25000],"duration":50,"cooldown":20,"energy":100},"13":{"className":"operator","level":[10,11,12,14,22],"cooldown":100,"duration":300,"range":3,"effect":[50,100,150,200,250],"period":15},"14":{"className":"operator","level":[10,11,12,14,22],"cooldown":100,"duration":100,"range":3,"effect":[2,4,6,8,10],"period":10},"15":{"className":"operator","level":[20,21,22,23,24],"cooldown":8,"duration":10,"range":50,"ops":[50,40,30,20,10]},"16":{"className":"operator","level":[10,11,12,14,22],"cooldown":800,"range":3,"duration":1000,"ops":200,"effect":[1,2,3,4,5]},"17":{"className":"operator","level":[0,2,7,14,22],"cooldown":5,"range":3,"ops":5,"duration":[1,2,3,4,5]},"18":{"className":"operator","level":[20,21,22,23,24],"cooldown":800,"range":3,"duration":1000,"ops":200,"effect":[10,20,30,40,50]},"19":{"className":"operator","level":[0,2,7,14,22],"cooldown":1000,"range":3,"duration":800,"ops":100}};
/*{
    [PWR_GENERATE_OPS]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 50,
        effect: [1, 2, 4, 6, 8]
    },
    [PWR_OPERATE_SPAWN]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 300,
        duration: 1000,
        range: 3,
        ops: 100,
        effect: [0.9, 0.7, 0.5, 0.35, 0.2]
    },
    [PWR_OPERATE_TOWER]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 10,
        duration: 100,
        range: 3,
        ops: 10,
        effect: [1.1, 1.2, 1.3, 1.4, 1.5]
    },
    [PWR_OPERATE_STORAGE]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 800,
        duration: 1000,
        range: 3,
        ops: 100,
        effect: [500000,1000000,2000000,4000000,7000000]
    },
    [PWR_OPERATE_LAB]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 50,
        duration: 1000,
        range: 3,
        ops: 10,
        effect: [2,4,6,8,10]
    },
    [PWR_OPERATE_EXTENSION]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 50,
        range: 3,
        ops: 2,
        effect: [0.2, 0.4, 0.6, 0.8, 1.0]
    },
    [PWR_OPERATE_OBSERVER]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 400,
        duration: [200,400,600,800,1000],
        range: 3,
        ops: 10,
    },
    [PWR_OPERATE_TERMINAL]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 500,
        duration: 1000,
        range: 3,
        ops: 100,
        effect: [0.9, 0.8, 0.7, 0.6, 0.5]
    },
    [PWR_DISRUPT_SPAWN]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 5,
        range: 20,
        ops: 10,
        duration: [1,2,3,4,5]
    },
    [PWR_DISRUPT_TOWER]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 0,
        duration: 5,
        range: 3,
        ops: 10,
        effect: [0.9, 0.8, 0.7, 0.6, 0.5],
    },
    [PWR_DISRUPT_SOURCE]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 100,
        range: 3,
        ops: 100,
        duration: [100, 200, 300, 400, 500]
    },
    [PWR_SHIELD]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        effect: [5000, 10000, 15000, 20000, 25000],
        duration: 50,
        cooldown: 20,
        energy: 100,
    },
    [PWR_REGEN_SOURCE]: {
        className: POWER_CLASS.OPERATOR,
        level: [10, 11, 12, 14, 22],
        cooldown: 100,
        duration: 300,
        range: 3,
        effect: [50,100,150,200,250],
        period: 15
    },
    [PWR_REGEN_MINERAL]: {
        className: POWER_CLASS.OPERATOR,
        level: [10, 11, 12, 14, 22],
        cooldown: 100,
        duration: 100,
        range: 3,
        effect: [2,4,6,8,10],
        period: 10
    },
    [PWR_DISRUPT_TERMINAL]: {
        className: POWER_CLASS.OPERATOR,
        level: [20, 21, 22, 23, 24],
        cooldown: 8,
        duration: 10,
        range: 50,
        ops: [50,40,30,20,10]

    },
    [PWR_FORTIFY]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 5,
        range: 3,
        ops: 5,
        duration: [1, 2, 3, 4, 5]
    },
    [PWR_OPERATE_POWER]: {
        className: POWER_CLASS.OPERATOR,
        level: [10, 11, 12, 14, 22],
        cooldown: 1000,
        range: 3,
        duration: 800,
        ops: 200,
        effect: [1, 2, 3, 4, 5]
    },
    [PWR_OPERATE_CONTROLLER]: {
        className: POWER_CLASS.OPERATOR,
        level: [20, 21, 22, 23, 24],
        cooldown: 1000,
        range: 3,
        duration: 800,
        ops: 200,
        effect: [10, 20, 30, 40, 50]
    },
    [PWR_OPERATE_FACTORY]: {
        className: POWER_CLASS.OPERATOR,
        level: [0, 2, 7, 14, 22],
        cooldown: 1000,
        range: 3,
        duration: 800,
        ops: 100
    }
};*/